File: src/algorithms/python/binary-search.py


def binary_search(array, first, last, item, steps=0):
    """A recursive binary search.

    :param list array: a sorted list to search
    :param int first: the index of the first item. During the first
        iteration this will be set to 0 but may change during the search
    :param int last: the index of the last element. During the first
        iteration this will be set to n-1 but may change during the search
    :param int item: the item to look for. This binary search assumes the
        list is an integer/float list.
    :param int steps: the number of recursive steps the search took.
        Purely for statistical uses.

    :returns: the index of the item or -1, and the number of steps it took to
        do the search.
    :rtype: tuple(int, int)
    """

    steps += 1
    middle = (last + first) // 2

    if last >= first:

        if array[middle] == item:
            return middle, steps

        if item > array[middle]:
            return binary_search(array, middle + 1, last, item, steps=steps)
        else:
            return binary_search(array, first, middle - 1, item, steps=steps)

    else:
        return -1, steps


# ------------------- tests ----------------

x = range(1049)

assert binary_search(x, 0, len(x) - 1, 256)[0] != -1
assert binary_search(x, 0, len(x) - 1, 1)[0] != -1
assert binary_search(x, 0, len(x) - 1, 1048)[0] != -1
assert binary_search(x, 0, len(x) - 1, 0)[0] != -1
assert binary_search(x, 0, len(x) - 1, 100)[0] != -1
assert binary_search(x, 0, len(x) - 1, 750)[0] != -1
assert binary_search(x, 0, len(x) - 1, 512)[0] != -1
assert binary_search(x, 0, len(x) - 1, 1000)[0] != -1
assert binary_search(x, 0, len(x) - 1, 1050)[0] == -1
assert binary_search(x, 0, len(x) - 1, 2000)[0] == -1
assert binary_search(x, 0, len(x) - 1, -8)[0] == -1

# ---------------------


def find_col_peak(two_d_array, col):
    """Find a column peak in a given 2D list.

    :param two_d_array: a 2D list
    :type: list[list[int]]
    :param int col: the column number to seach

    :returns: a tuple of colum max index and column max value
    :rtype: tuple(int, int)
    """
    col_max_value = two_d_array[0][col]
    col_max = 0
    for row in range(1, len(two_d_array)):
        new_number = two_d_array[row][col]
        if new_number > col_max_value:
            col_max_value = new_number
            col_max = row

    return col_max, col_max_value


def peak_finder(two_d_array, first_column, last_column):
    """A recursive 2D peak finder.

    :param two_d_array: a 2D list
    :type: list[list[int]]
    :param int first_column: the first column of the matrix, intially set to 0
    :param int last_column: the last column of the matrix, initally set to n-1

    :return: a vector of x, y coordinates of the peak (column, row)
    :rtype: tuple(int, int)
    """

    middle_column = last_column // 2
    row, item = find_col_peak(two_d_array, middle_column)

    # if we are in the first or last row then that column max will ve a local
    # maximum
    if middle_column == first_column or middle_column == last_column:
        return middle_column, row

    if last_column > first_column:

        # if we find peak
        if (two_d_array[row][middle_column - 1]
            <= item >= two_d_array[row][middle_column + 1]):
            return middle_column, row

        if two_d_array[row][middle_column - 1] > item:
            return peak_finder(two_d_array, first_column, middle_column - 1)

        else:
            return peak_finder(two_d_array, middle_column + 1, last_column)

    return middle_column, row


arr = [
    [10, 8, 10, 10, 16],
    [14, 13, 12, 11, 24],
    [15, 9, 11, 21, 6],
    [16, 21, 19, 20, 12],
    [100, 6, 13, 19, 9]
]

print(peak_finder(arr, 0, len(arr) - 1))

File: src/algorithms/c/binarySearch.c
#include <assert.h>
#include <stdio.h>
#include <stdlib.h>

typedef struct max_vector {
    unsigned int col_index;
    unsigned int row_index;
    int index_value;
} max_vector;


int binary_search(int arr[], int first, int last, int item) {
    /* A recursive binary search */

    int middle = (last + first) / 2;
    if (last >= first) {

        if (arr[middle] == item) {
            return middle;
        }

        if (item > arr[middle]){
            return binary_search(arr, middle + 1, last, item);
        } else {
            return binary_search(arr, first, middle - 1, item);
        }
  
    } else {
        return -1;
    }
}


max_vector find_col_peak(int *arr, int rows, int col){
    /* Find the peak of a column.
    
    This function treats a 2D array like a 1D array and
    traverses using some basic pointer arithmetic.

    next item = *((i * row + arr) + j)
    where:
        - i == current row number (the y axis in matrix i.e. up and down)
        - j == column number in column space (the x axis i.e left to right)
        - row == the number of items in each row
        - *arr == the first item in the array
    */

    // we want a vector to save the max number and its index
    max_vector vector;
    
    // set max to the first item of the column
    vector.col_index = 0;
    vector.index_value = *(col + arr);

    // loop through the rest of the column
    for (int i=1; i < rows; i++){
        int next_number = arr[i * rows + col];
        if (next_number > vector.index_value){
            vector.col_index = i;
            vector.index_value = next_number;
        }
    }
    printf("the max value is %d and it is at index %d\n",
        vector.index_value, vector.col_index);
    return vector;
}


max_vector peak_finder(int *matrix, int rows, int first, int last){
    /* Find a local 2D peak.

    NOTE: Expects a matrix of size row x col

    A peak is defined as:
    (i, j) >= (i, j + 1), (i, j - 1), (i + 1, j) (i - 1, j)
    */

    max_vector vector;
    
    int middle = last / 2;
    vector = find_col_peak(matrix, rows, middle);

    if (middle == first || middle == last){
        vector.row_index = middle;
        return vector;
    }

    if (last > first) {
        int current = matrix[vector.col_index * rows + middle];
        int one_less = matrix[vector.col_index * rows + middle - 1];
        int one_more = matrix[vector.col_index * rows + middle + 1];

        if (one_less <= current >= one_more) {
            vector.row_index = middle;
            return vector;
        }

        // if current is not a peak we can free the vector
        if (one_less > current){
            return peak_finder(matrix, rows, first, middle - 1);
        } else {
            return peak_finder(matrix, rows, middle + 1, last);
        } 
    }
    vector.row_index = middle;
    return vector;
}


int main(void) {
    printf("Hello World\n");

    int myarr[10] = {1, 2, 3, 4, 5, 6, 7, 8, 9, 10};

    // --------- binary search tests -------
    assert(binary_search(myarr, 0, 10, 5) == 4);
    assert(binary_search(myarr, 0, 10, 1) == 0);
    assert(binary_search(myarr, 0, 10, 10) == 9);
    assert(binary_search(myarr, 0, 10, 0) == -1);
    assert(binary_search(myarr, 0, 10, -2) == -1);

    // ------------ 2D peak ----------

    int arr [5][5] = {
        {10, 8, 11, 10, 16},
        {14, 13, 12, 11, 24},
        {15, 9, 11, 21, 6},
        {16, 21, 19, 20, 12},
        {100, 6, 13, 19, 9},
    };

    max_vector ptr = peak_finder(*arr, 5, 0, 5);
    printf("the local peak is at %d, %d", ptr.row_index, ptr.col_index);
    return 0;
}
